text/0000-serialize-deserialize.md

```diff
@@ -0,0 +1,174 @@
+token. 
+
+Providing a standard API for sharing json data between server and browser plugins
+will make it easier to test these types of plugins in isolation without needing to resort 
+to integration tests. Additionally, it can help solve problems where users may want custom 
+serialization / deserialization mechanisms, such as using tools like immutable.js.
+
+# Detailed design
+
+This proposal adds the following APIs to `fusion-core` 
+
+### `SharedStateToken`
+
+```js
+import {SharedStateToken} from 'fusion-core';
+```
+
+The `SharedStateToken` will have a default implementation registered onto each fusion app
+which provides a `{serialize}` and `{deserialize}` function which serialize json data on the
+server into a `<script type='application/json'>` dom element which is appended to `ctx.template.body`, and
+deserialize the data from the dom node in the browser. Example Usage:
+
+```js
+import {SharedStateToken, createPlugin} from 'fusion-core';
+
+export default createPlugin({
+  deps: {
+    SharedState: SharedStateToken,
+  },
+  middleware: ({SharedState}) => {
+    const {serialize, deserialize} = SharedState('some-id');
+    return (ctx, next) => {
+      if (__BROWSER__) {
+        // load and parse serialized data from dom node
+        const data = deserialize();
+      } else {
+        // serialize json data into body
+        serialize(ctx, {
+          some: 'json',
+        });
+      }
+      return next();
+    };
+  },
+});
+```
+
+### `MockSharedState`
+
+```js
+import {MockSharedState} from 'fusion-core';
+```
+
+This API will provide an easy way to provide mock data for all deserialize
+calls in an application. 
+
+```js
+import {MockSharedState, SharedStateToken} from 'fusion-core';
+// Deserialize calls will now return the mock values for these ids
+app.register(SharedStateToken, MockSharedState({
+  'some-id': {
+    some: 'json'
+  },
+  'other-data': {}
+}));
+```
+
+### Custom Serialization
+
+This structure can also be useful when trying to implement custom serialization / deserialization for certain plugins. 
+For example, lets say you want to use the `fusion-plugin-react-redux` plugin but want to have custom serialization / deserialization
+to support immutable-js in your store. You can create your own custom serializer/deserializer, and even set up your app to only
+use this custom SharedState implementation for the react redux plugin using token aliasing.
+
+```js
+import {createToken, html, SharedStateToken} from 'fusion-core';
+import ReduxPlugin, {ReduxToken} from 'fusion-plugin-react-redux';
+import transit from 'transit-immutable-js';
+
+const ImmutableSharedState = (id) => {
+  return {
+    serialize: (ctx, data) => {
+      ctx.template.body.push(
+        html`<div id='${id}'>${transit.toJSON(data)}</div>`
+      );
+    },
+    deserialize: () => {
+      return transit.fromJSON(document.getElementById(id).textContent);
+    }
+  }
+}
```

Comment:
In this example, wouldn't it be better to just embed the serialized `immutable-transit` strings into the standard JSON serializer/deserializer (rather than a custom out-of-band serialization)?

I can't think of anything that would need for *true* custom serialization aside from CSS-in-JS implementations.

Reply:
I think the actual HTML part of the serialization and standardization can be kept in core. Custom deserializers/serializers need not involve HTML at all, just `A → string` and `string → A` functions, and the core SSR plugin itself can handle the HTML part.

So you'd have something sort of like this.

```js
const DefaultSharedState = {
  serialize: data => JSON.stringify(data),
  deserialize: serializedData => JSON.parse(serializedData)
};

const ImmutableSharedState = {
  serialize: data => transit.toJSON(data),
  deserialize: serializedData => transit.fromJSON(serializedData)
};
```

I think the need for anything to actually have direct control of HTML (rather than just string serialization/deserialization) is only for the special case of CSS-in-JS, and in that case, they will want to implement that in a totally custom way rather than using this.

Reply:
I think it's doubtful that constraint will be an issue in practice. The only non-contrived use case I've been able to come up thus far is CSS-in-JS, but it's unclear to me how mocking state would make sense in that case.

Reply:
ok - I can update the RFC to reflect that API

Reply:
Potentially what might be simpler is all values are always `JSON.stringify`/`JSON.parse`, and if you want something else you can just serialize/deserialize into a string, which is valid JSON. Although I suppose this might be more overhead.